src/reboost/build_elm.py

```diff
@@ -0,0 +1,264 @@
+    lh5_table
+        the table name to read.
+    stp_file
+        the file name path.
+    id_name
+        the name of the `evtid` field.
+    start_row
+        the row to begin reading.
+    last_vertex_evtid
+        the last evtid to read up to.
+    stp_buffer
+        the number of rows to read at once.
+
+    Returns
+    -------
+         a tuple of the updated `start_row`, the first row for the chunk and an awkward Array of the steps.
+    """
+
+    # make a LH5Store
+    store = LH5Store()
+
+    # some outputs
+    evtids_proc = None
+    last_evtid = 0
+    chunk_start = 0
+
+    # get the total number of rows
+    n_rows_tot = store.read_n_rows(f"{lh5_table}/{id_name}", stp_file)
+
+    # iterate over the file
+    # stop when the entire file is read
+
+    while start_row < n_rows_tot:
+        # read the file
+        lh5_obj, n_read = store.read(
+            f"{lh5_table}/{id_name}",
+            stp_file,
+            start_row=start_row,
+            n_rows=stp_buffer,
+        )
+        evtids = lh5_obj.view_as("ak")
+
+        # if the evtids_proc is not set then this is the first valid chunk
+        if evtids_proc is None:
+            evtids_proc = evtids
+            chunk_start = start_row
+        elif evtids[0] <= last_vertex_evtid:
+            evtids_proc = ak.concatenate((evtids_proc, evtids))
+
+        # get the last evtid
+        last_evtid = evtids[-1]
+
+        # if the last evtid is greater than the last vertex we can stop reading
+        if last_evtid > last_vertex_evtid or (start_row + n_read >= n_rows_tot):
+            break
+
+        # increase rhe start row for the next read
+
+        if start_row + n_read <= n_rows_tot:
+            start_row += n_read
+
+    return start_row, chunk_start, evtids_proc
+
+
+def build_elm(
+    stp_file: str,
+    elm_file: str | None,
+    *,
+    id_name: str = "g4_evtid",
+    evtid_buffer: int = 10000,
+    stp_buffer: int = 1000,
+) -> ak.Array | None:
+    """Builds a g4_evtid look up (elm) from the stp data.
+
+    This object is used by `reboost` to efficiency iterate through the data.
+    It consists of a :class:`LGDO.VectorOfVectors` for each lh5_table in the input files.
+    The rows of this :class:`LGDO.VectorOfVectors` correspond to the `id_name` while the data
+    are the `stp` indices for this event.
+
+    Parameters
+    ----------
+    stp_file
+        path to the stp (input) file.
+    elm_file
+        path to the elm data, can also be `None` in which case an `ak.Array` is returned in memory.
+    id_name
+        name of the evtid file, default `g4_evtid`.
+    stp_buffer
+        the number of rows of the step file to read at a time
+    evtid_buffer
+        the number of evtids to read at a time
+
+    Returns
+    -------
+    either `None` or an `ak.Array`
+    """
+
+    store = LH5Store()
+
+    # loop over the lh5_tables
+    lh5_table_list = [table for table in lh5.ls(stp_file, "stp/") if "vertices" not in table]
```

Comment:
`if "vertices" not in table`

this check sounds unstable (I cannot think of a reason to have a "vertices" column in _any_ other table, but nevertheless we should use the right checks...)